### ISSUE TITLE
`npx prisma migrate dev --create-only` actually resets database

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#8053]
It tells you it will do it:
```
...
√ Drift detected: Your database schema is not in sync with your migration history.

[+] Added tables
  - Buyer
  - BuyerGroup

[*] Changed the `Buyer` table
  [+] Added foreign key on columns (buyerGroupId)


We need to reset the PostgreSQL database "emerald_jellyfish" at "pg-db-provision.cm0mkpwj8arx.eu-central-1.rds.amazonaws.com:5432".
Do you want to continue? All data will be lost. ... yes
...
```

But in the end it is still surprising that a command that is supposed to only create a migration (file), modifies the database.

**Comment by tomhoule** [comment]
One more issue to file into "works as intended, but counter-intuitive".

**Comment by deemeetree** [comment]
In fact, in the end it provides this message, but it doesn't remove the data. Although the migration file that it creates will remove the data if ran using `prisma migrate dev`. The question is: how to avoid it dropping the tables if I just want to update them?

**Comment by floelhoeffel** [comment]
@deemeetree `migrate dev` enforces a clean state (migration history in file system and development database need to be in sync) to calculate the next migration. There is 3 use cases where this could not be the case when running `prisma migrate dev --create-only`:

1. drift was detected (the development db is ahead of the migration history in the file system)
2. migration files were changed after they were applied (detected by checksum comparison)
3. unapplied migrations are found in the filesystem

It's unintuitive that a command which says "createOnly" actually does more than "create" as it needs to create a clean state first.

There is a good summary in our [engines faq](https://github.com/prisma/prisma-engines/blob/main/migration-engine/ARCHITECTURE.md#why-does-migrate-dev-constantly-want-to-reset-my-database-could-it-not-simply-roll-back-to-the-desired-state)

**Comment by romgrk-comparative** [comment]
@floelhoeffel is there are way to find which of these option applies? I want to add a new field to my database, but the migration command really wants to reset my database and I don't understand why.

**Comment by floelhoeffel** [comment]
@romgrk-comparative you could try our new diffing tool to find where the difference sits. 

https://www.prisma.io/docs/reference/api-reference/command-reference#migrate-diff

**Comment by tomhoule** [comment]
@romgrk-comparative what does the command print? We strive to always provide a clear reason for resets, if the output is not informative enough, we'd consider that a bug.

**Comment by romgrk-comparative** [comment]
I can't test the diffing tool anymore but will do on the next occasion I have.

What's weird is, once the migration file was created, `npx prisma migrate deploy` had no issue applying the migration without resetting the (production) database. It really only wanted to reset the database to create the migration file. I chose to say ok for the reset using the development database, for which I have reset/repopulate scripts.

**Comment by a-eid** [comment]
I've been using `prisma db push` whenever I make a change to the DB, the current state of the DB / Schema right now is what I want.

how do I avoid resetting / losing data when creating a migration ?

**Comment by janpio** [comment]
Yes. You want to use baselining instead: https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/baselining

**Comment by a-eid** [comment]
@janpio thinking about it, what if I don't need the previous migrations, can I remove all migrations and start fresh without losing the data in the database ?

**Comment by janpio** [comment]
You don't have any previous "migrations" (= migration files) when you used `db push`. That is exactly what baselining is needed to fix - so that you have one migration that brings your migration history in a state that reflects your current database state. That will allow you to keep the current database state, including the data.

**Comment by a-eid** [comment]
@janpio I do, I used to use `migrations` until I learned about `prisma db push`.

<img width="333" alt="Screenshot 2023-01-06 at 23 42 11" src="https://user-images.githubusercontent.com/19273413/211104860-0094b517-e193-4b69-aa89-4f6715db051e.png">

**Comment by janpio** [comment]
That would indeed work @a-eid - but optimally Prisma could offer an adapted baseline command here. You really need something to create a "migration.sql" that describes how to go from where the migration history currently brings you - to what you have right now.

**Comment by Thinkscape** [comment]
This is extremely annoying.
Even though the diff sees incremental changes to the schema, it forces me to reset the whole thing for no clear reason.

I didn't find any workaround or command to generate migration without nuking the connected database 😮‍💨

**Comment by janpio** [comment]
That sounds like a valid suggestion.
We will probably only be able to fully validate that when it exists and one can play with it though, so this would be a high risk addition that might not get merged in the end.

**Comment by trulyronak** [comment]
this is one of the biggest frustrations with prisma
I've only been doing `prisma migrate dev --create-only` to create migration files (followed by `prisma migrate dev` to apply said migration), but occasionally I find myself in this situation, where the `create-only` claims database drift. There isn't any!  If I blank reset it, migrate deploy has no issues.

What's going on?

**Comment by trulyronak** [comment]
Running `pnpm dlx prisma migrate status`, it says I haven't run any of the migrations → but i was → what's the best option here?

**Comment by janpio** [comment]
Please open a new issue for your specific problem @trulyronak, it sounds like your `migrate dev` is detecting drift where it should not - that is a separate problem we should and want to look into. We'll need as much information as possible to hopefully be able to reproduce and understand this.
